src/menu_helpers.c

```diff
@@ -0,0 +1,478 @@
+#include "items.h"
+#include "link.h"
+#include "mail_data.h"
+#include "main.h"
+#include "map_constants.h"
+#include "menu.h"
+#include "menu_helpers.h"
+#include "rom4.h"
+#include "songs.h"
+#include "sound.h"
+#include "sprite.h"
+#include "task.h"
+#include "text.h"
+
+static EWRAM_DATA u8 gUnknown_020388C0[4] = {0};
+static EWRAM_DATA struct YesNoFuncTable gUnknown_020388C4 = {0};
+
+static TaskFunc gUnknown_0300074C;
+
+extern struct SpritePalette gUnknown_083E5968;
+
+extern struct SpriteTemplate gSpriteTemplate_83E59D0;
+extern struct SpriteTemplate gSpriteTemplate_83E59E8;
+extern struct SpriteTemplate gSpriteTemplate_83E5A00;
+
+extern const u8 gUnknown_083E5A18[];
+extern const u8 gUnknown_083E5A1C[];
+
+void sub_80F9020(void)
+{
+    memset(&gBGTilemapBuffers[1], 0, 0x800);
+    memset(&gBGTilemapBuffers[2], 0, 0x800);
+    memset(&gBGTilemapBuffers[3], 0, 0x800);
+}
+
+// display message box, fill box with tile if tile is not zero, print string
+static void PrintMessage(const u8 *str, u16 tile)
+{
+    MenuDisplayMessageBox();
+    if (tile)
+    {
+        sub_80A3FA0(&gBGTilemapBuffers[1][0], 2, 15, 26, 4, tile);
+    }
+    sub_8072044(str);
+}
+
+static void sub_80F9090(u8 taskId)
+{
+    if (MenuUpdateWindowText() == TRUE)
+    {
+        gUnknown_0300074C(taskId);
+    }
+}
+
+void DisplayItemMessageOnField(u8 taskId, const u8 *str, TaskFunc callback, u16 tile)
+{
+    PrintMessage(str, tile);
+    gUnknown_0300074C = callback;
+    gTasks[taskId].func = sub_80F9090;
+}
+
+static void Task_CallYesOrNoCallback(u8 taskId)
+{
+    switch (ProcessMenuInputNoWrap_())
+    {
+    case 0:
+        PlaySE(SE_SELECT);
+        gUnknown_020388C4.yesFunc(taskId);
+        break;
+    case 1:
+    case -1:
+        PlaySE(SE_SELECT);
+        gUnknown_020388C4.noFunc(taskId);
+    }
+}
+
+void DoYesNoFuncWithChoice(u8 taskId, const struct YesNoFuncTable *funcTable)
+{
+    gUnknown_020388C4 = *funcTable;
+    gTasks[taskId].func = Task_CallYesOrNoCallback;
+}
+
+static void PrintStringWithPalette(const u8 *str, u8 paletteNum, u8 left, u8 top)
+{
+    u8 paletteStr[4];
+
+    memcpy(paletteStr, gUnknown_083E5A18, 4);
+
+    if (paletteNum != 0xFF)
+    {
+        paletteStr[2] = paletteNum;
+        StringCopy(gStringVar4, paletteStr);
+        StringAppend(gStringVar4, str);
+        paletteStr[2] = gMenuWindowPtr->paletteNum;
+        StringAppend(gStringVar4, paletteStr);
+    }
+    else
+    {
+        StringCopy(gStringVar4, str);
+    }
+
+    MenuPrint(gStringVar4, left, top);
+}
+
+// unused
+void PrintNumberWithPalette(s32 value, u8 paletteNum, u8 n, u8 mode, u8 left, u8 top)
+{
+    ConvertIntToDecimalStringN(gStringVar1, value, mode, n);
+    if (paletteNum != 0xFF)
+    {
+        PrintStringWithPalette(gStringVar1, paletteNum, left, top);
+    }
+    else
+    {
+        MenuPrint(gStringVar1, left, top);
+    }
+}
+
+// unused
+void PrintTriangleCursorWithPalette(u8 left, u8 top, u8 paletteNum)
+{
+    u8 buffer[2];
```

Comment:
These `memcpy`s should be fairly trivial to replace with `u8 buffer[] = _("▶$")`; That should place the string in the `.rodata` section of the object file.